// @egjs/ngx-flicking/esm2020/lib/ngx-flicking-panel.directive.mjs
/*
 * Copyright (c) 2015 NAVER Corp.
 * egjs projects are licensed under the MIT license
 */
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class NgxFlickingPanel {
    get nativeElement() {
        return this._host.nativeElement;
    }
    get rendered() {
        return this._rendered;
    }
    constructor(_host, _renderer) {
        this._host = _host;
        this._renderer = _renderer;
        this._rendered = true;
    }
    show(flicking) {
        this._rendered = true;
        const el = this.nativeElement;
        const cameraEl = flicking.camera.element;
        if (el.parentElement !== cameraEl) {
            this._renderer.appendChild(cameraEl, el);
        }
    }
    hide(flicking) {
        this._rendered = false;
        const el = this.nativeElement;
        const cameraEl = flicking.camera.element;
        if (el.parentElement === cameraEl) {
            this._renderer.removeChild(cameraEl, el);
        }
    }
}
NgxFlickingPanel.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: NgxFlickingPanel, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
NgxFlickingPanel.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.2", type: NgxFlickingPanel, isStandalone: true, selector: "[flicking-panel], [FlickingPanel]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: NgxFlickingPanel, decorators: [{
            type: Directive,
            args: [{
                    selector: '[flicking-panel], [FlickingPanel]',
                    standalone: true,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; } });
